### PR TITLE
Adabooststump algorithm change

### PR DESCRIPTION
**Pull request by yufongpeng**
Originally, the coeff for each stump was 1/2 * log((1+err)/(1-err)). It was positively correlated to the error rate, which was kind of odd. I change the coeff into log((1-err)/err) + log(n_labels - 1), called SAMME (https://www.intlpress.com/site/pub/pages/journals/items/sii/content/vols/0002/0003/a008/), and it was one of the algorithm that scikitlearn uses (another one is SAMME.R, which has better performance). I've tested the performance; mostly, the SAMME is comparable to the old one.  However , I created a mini test:
```
julia> x
10×2 Matrix{Int64}:
 1  4
 1  3
 1  4
 1  3
 2  2
 2  4
 2  6
 2  4
 2  5
 2  6

julia> y
10-element Vector{Int64}:
 1
 1
 1
 1
 1
 2
 2
 2
 2
 2
```
The first feature can classify most observation but the fifth one need to be discriminated by the second feature
With the old method:
```
julia> map(1:8) do i
       model, coeffs = DecisionTree.build_adaboost_stumps(y, x, i)
       preds = apply_adaboost_stumps(model, coeffs, x);
       cm = confusion_matrix(y, preds)
       cm.accuracy
       end
8-element Vector{Float64}:
 0.9
 0.9
 0.9
 0.9
 0.9
 0.9
 1.0
 1.0
```
With the SAMME algorithm:
```
julia> map(1:8) do i
       model, coeffs = DecisionTree.build_adaboost_stumps(y, x, i)
       preds = apply_adaboost_stumps(model, coeffs, x);
       cm = confusion_matrix(y, preds)
       cm.accuracy
       end
8-element Vector{Float64}:
 0.9
 0.9
 1.0
 1.0
 1.0
 1.0
 1.0
 1.0
```
This is not a practical use and I've tested other larger data set like digits and iris, there's no difference on converging rate.

### CODE REVIEW

**Comment by ablaom** [comment]
@bensadeghi Be great to get your input on this particular issue.

**Comment by bensadeghi** [comment]
I don't recall where I got the current formula for the coefficients, but the SAMME approach looks like. 
Let's just make sure that these new weights deliver good cross-validation results, comparable to or better than the existing process.

**Comment by ablaom** [comment]
> Let's just make sure that these new weights deliver good cross-validation results, comparable to or better than the existing process.

@yufongpeng Would you be happy to check this? 

 And to look into the 1.6 fail?

**Comment by yufongpeng** [comment]
Here's the result from tests.
|datasets|n_iterations|n_folds|old methods|new methods|
|--------|------------|-------|-----------|-----------|
|adults|15|3|0.839|0.844|
|iris|15|3|0.933|0.973|
|low presision|25|3|0.868|0.895|
|random(rng=5)|25|3|0.874|0.888|
|random(rng=10)|25|3|0.889|0.898|

1.6 fail is because `_nfoldCV` doesn't pass `rng` argument into `build_adabooststumps`. I'll fix this as well.

**Comment by ablaom** [comment]
@yufongpeng Thanks for that! What metric are you using? Assuming this is a score and not a loss this looks fine,  although as a sanity check, it would be good to include the standard errors if these are available.

**Comment by yufongpeng** [comment]
Here is updated tables.
The metric is accuracy ± standard error
|datasets|n_iterations|n_folds|old method|new method|
|--------|------------|-------|-----------|-----------|
|adults|15|3|0.829 ± 0.003|0.844 ± 0.006|
|iris|15|3|0.933 ± 0.064|0.967 ± 0.031|
|low presision|25|3|0.865 ± 0.006|0.891 ± 0.003|
|random(rng=5)|25|3|0.872 ± 0.011|0.888 ± 0.018|
|random(rng=10)|25|3|0.879 ± 0.009|0.895 ± 0.003|

**Comment by ablaom** [comment]
Great. I'm happy with changing the algorithm. Todo:

- [ ] Fix 1.6 fails
- [ ] Document the change, with link to SAMME paper. Since we don't have doc strings yet, it could go at line 22 of [readme](https://github.com/JuliaAI/DecisionTree.jl/edit/dev/README.md).

File: src/measures.jl [hunk 1]
```diff
@@ -135,7 +135,7 @@ function _nfoldCV(classifier::Symbol, labels::AbstractVector{T}, features::Abstr
             predictions = apply_forest(model, test_features)
         elseif classifier == :stumps
             model, coeffs = build_adaboost_stumps(
-                train_labels, train_features, n_iterations)
+                train_labels, train_features, n_iterations, rng=rng)
```
Comment (ablaom) [inline src/measures.jl:138]: Great catch!

Comment (ablaom) [reply src/measures.jl:138]: Please see https://github.com/JuliaAI/DecisionTree.jl/pull/174#issuecomment-1159931376 .

**Review by ablaom — APPROVED** [review]
Approved. But as breaking,  waiting for:

- [x] Separate patch release to address rng bug [discovered in this](https://github.com/JuliaAI/DecisionTree.jl/pull/167/files#r901224499) PR. See comments at #174

**Comment by ablaom** [comment]
@yufongpeng The RNG fixes at #174 (which include the one here) has been merged. And I'm not planning any more pre v0.11 PRs.  Can you please resolve merge conflicts, ensuring  [@rikhuijzer's comment](https://github.com/JuliaAI/DecisionTree.jl/pull/167#discussion_r901352148) is also addressed? The after #178, we should be ready to roll.

**Comment by yufongpeng** [comment]
@ablaom I've resolved the issues, and the PR should be ready to go.

**Comment by ablaom** [comment]
Great @yufongpeng . Will merge shortly. Thanks for your patience.

**Pull request merged**